[TS] Part 3: Targeted Sweep Queue Progress Table + Class

gmaretic opened the pull request:
**Goals (and why)**:
Implement the progress table for sweep queue.

**Implementation Description (bullets)**:
The table stores the last swept timestamp for each shard. It also stores the information about the number of shards we are using.

**Concerns (what feedback would you like?)**:
Just using a row of this table to store the number of shard information is hacky, but does not seem worth it to have a separate table just for that. Alternatively, we could create the new rows eagerly, but that is potentially a lot of CAS operations at once.

Also, the treatment of edge cases with CAS failing etc is open to discussion.

**Where should we start reviewing?**:
`KvsSweepProgressQueueTest`

**Priority (whenever / two weeks / yesterday)**:
Not urgent, but small (ignoring generated code)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/palantir/atlasdb/3108)
<!-- Reviewable:end -->

gsheasby commented on atlasdb-client/src/main/java/com/palantir/atlasdb/schema/TargetedSweepSchema.java:
```diff
@@ -75,6 +75,19 @@ private static Schema generateSchema() {
+                rowComponent("sweep_conservative", ValueType.BLOB);
+            columns();
+                column("value", "v", ValueType.VAR_LONG);
+            sweepStrategy(TableMetadataPersistence.SweepStrategy.NOTHING);
```
It seems to me like we _would_ want to sweep this table.

gsheasby replied:
Upon re-reading the design doc: or maybe cleaning these tables up will come later?

gsheasby commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/KvsSweepQueueProgress.java:
```diff
@@ -0,0 +1,116 @@
+public class KvsSweepQueueProgress {
+    private static final TableReference TABLE_REF = TargetedSweepTableFactory.of()
+            .getSweepShardProgressTable(null).getTableRef();
+    private static final int NUMBER_OF_SHARDS_INDEX = 1024;
```
name could be clearer. Do you mean "number of shards in index"? Or something else? `SHARDS_IN_INDEX`?

gsheasby replied:
Oh, this is a special index for saying how many shards there are? -_-

If we must do that rather than have a new table, can we use `-1`? `SHARD_COUNT_INDEX`?

jeremyk-91 replied:
I like `SHARD_COUNT_INDEX`.

Heads up: if you change the magic value to `-1`, then you'll need to declare the typing as `VAR_SIGNED_LONG`

gsheasby commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/KvsSweepQueueProgress.java:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    public long numberOfShards() {
+        return getOrInitializeTo(NUMBER_OF_SHARDS_INDEX, true, 1L);
```
Strong suggestion: use an enum rather than bool, for readability and to reduce mental mapping.

gsheasby commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/KvsSweepQueueProgress.java:
```diff
@@ -0,0 +1,116 @@
+        increaseValue(NUMBER_OF_SHARDS_INDEX, true, newNumber);
+    }
+
+    public long lastSweptTimestamp(int shard, boolean conservative) {
```
I know it's out of fashion to use verbs, but `getLastSweptTimestamp` would be clearer here.

gmaretic replied:
Done in future PR

gsheasby commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/KvsSweepQueueProgress.java:
```diff
@@ -0,0 +1,116 @@
+            kvs.checkAndSet(initializeRequest);
+            return initialVale;
+        } catch (CheckAndSetException e) {
+            return getValue(getEntry(shard, conservative));
```
Do we care enough about this case to log it?

gsheasby commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/KvsSweepQueueProgress.java:
```diff
@@ -0,0 +1,116 @@
+        long oldVal = getValue(getEntry(shard, conservative));
+        SweepShardProgressTable.Value colValNew = SweepShardProgressTable.Value.of(newValue);
+
+        while (oldVal < newValue) {
```
Again, do we care about the case where `oldVal` is already greater than `newValue`?

jeremyk-91 replied:
probably not. Maybe the method should be named `increaseToAtLeast` or something like that?

gsheasby commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/KvsSweepQueueProgress.java:
```diff
@@ -0,0 +1,116 @@
+        }
+    }
+
+    private void increaseValue(int shard, boolean conservative, long newValue) {
```
The value stored in the database after this call might be more than `newValue`. Do we want to return the last known value, so that any consequences of this can be handled by the caller? I can't tell this by looking at this class in isolation, but it's something to bear in mind as we continue building this.

gsheasby commented on atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/queue/KvsSweepProgressQueueTest.java:
```diff
@@ -0,0 +1,122 @@
+    public void attemptingToDecreaseNumberOfShardsIsNoop() {
+        progress.numberOfShards();
+        progress.updateNumberOfShards(64);
+        progress.updateNumberOfShards(32);
```
No exception in this case?

gmaretic replied:
Subject to change, depending on how we implement the control over this parameter. I can see throwing if we set it as an install config, but probably want to WARN and continue using the old value still.

gsheasby commented on atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/queue/KvsSweepProgressQueueTest.java:
```diff
@@ -0,0 +1,122 @@
+        progress.updateLastSweptTimestamp(10, true, 128L);
+        assertThat(progress.lastSweptTimestamp(10, false)).isEqualTo(0L);
+
+        progress.updateLastSweptTimestamp(10, false, 32L);
```
Does it make sense to have different conservative and thorough timestamps per shard?

jeremyk-91 replied:
Depends on the implementation - but we're already treating them as separate queues (so we have 2 queues per shard) based on the timestamps writer in part 2, so I think it makes sense.

jeremyk-91 commented on atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/queue/KvsSweepProgressQueueTest.java:
```diff
@@ -0,0 +1,122 @@
+
+    @Test
+    public void increasingNumberOfShardsAbove128Throws() {
+        progress.numberOfShards();
```
nit: this line seems a bit random; if it's only for setup then it could make sense being called in a `@Before`. Nothing major though

gmaretic replied:
Changed this on another branch, no longer necessary to get progress to initialize first

jeremyk-91 commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/KvsSweepQueueProgress.java:
```diff
@@ -0,0 +1,116 @@
+        return value.getValue();
+    }
+
+    private long initializeShard(int shard, boolean conservative, long initialVale) {
```
nit: `initialValue`

gmaretic commented:
Addressed the comments that have not already incidentally been fixed on other branches. If you're good with these changes and have no further comments, should be good to merge

gsheasby commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/KvsSweepQueueProgress.java:
```diff
@@ -85,32 +88,33 @@ private long getValue(Map<Cell, Value> entry) {
+                oldVal = getValue(getEntry(shardAndStrategy));
             }
         }
+        return  oldVal;
```
OCD nit: double space here

gsheasby commented on atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/queue/KvsSweepProgressQueueTest.java:
```diff
@@ -63,60 +70,60 @@ public void increasingNumberOfShardsAbove128Throws() {
     @Test
     public void canReadInitialSweptTimestamp() {
-        assertThat(progress.lastSweptTimestamp(10, true)).isEqualTo(0L);
+        assertThat(progress.lastSweptTimestamp(ShardAndStrategy.conservative(10))).isEqualTo(INITIAL_TIMESTAMP);
```
nit: `CONSERVATIVE_TEN`
